Add support for auto starting voice and video calls

lukelabonte opened the pull request:
## Overview

This PR enhances the DoximityDialerSDK with three new public APIs that give vendors more control over call initiation and installation detection:

New APIs:
- `startVoiceCall(_:)` - Automatically initiates a voice call without user selection.
- `startVideoCall(_:)` - Automatically initiates a video call without user selection.
- `isDoximityInstalled` - Checks if the Doximity app is installed on the device, this already existed just wasn't public. 

Added an example app:
A comprehensive app that demonstrates all SDK features in three implementations:
- SwiftUI
- UIKit (Swift)
- UIKit (Objective-C)

## Technical Detail(s)

- All new methods include full Objective-C support via bridging. The existing dialPhoneNumber(_:) method continues to work as before (prefill mode where the user chooses call type).
- Updated App Store redirect URL from AppsFlyer to Singular
- Migrated tests to Swift Testing framework
- Enhanced documentation with usage examples and warnings

## Breaking Changes

None. All changes are additive. The existing `dialPhoneNumber(_:)` API remains unchanged and fully backward compatible.

## QA Recommendation

Developer QA (not traditional QA) because:

1. SDK/Library nature - This is a developer-facing SDK that requires integration into an app, not a standalone application
2. Example app is the test harness - The included example app demonstrates all functionality and serves as the primary validation tool

Suggested Developer QA Checklist:

- [x] Run SwiftUI example and test all three call methods
- [x] Run UIKit (Swift) example and test all three call methods
- [x] Run UIKit (Objective-C) example and test all three call methods
- [x] Verify installation detection badge shows correct status
- [x] Test with Doximity app installed - verify calls open Doximity correctly
- [x] Test without Doximity app - verify App Store redirect works

lukelabonte committed:
Fixed typo

lukelabonte committed:
Refactor URL handling to accept optional URLs

- Move guard statements from call sites into openURL helper method
- Change openURL parameter to accept URL? instead of URL
- Simplify dialPhoneNumber by removing redundant guard statements

lukelabonte committed:
Add support for auto-starting voice and video calls

- Add Options enum with prefill and startCall(Kind) cases
- Add Kind enum for voice and video call types
- Refactor URL construction to use URLComponents with dynamic paths
- Update dialPhoneNumber to accept options parameter (defaults to .prefill)
- Add comprehensive documentation for Options and Kind enums

lukelabonte committed:
Improve public API documentation

- Add documentation for DoximityDialerError enum and its cases
- Enhance dialPhoneNumber documentation with clear behavior descriptions
- Document throws and returns for icon methods
- Add comprehensive documentation for Objective-C bridge methods
- Include usage examples in parameter descriptions

lukelabonte committed:
Add Objective-C support for auto-starting calls

- Add method overload to support voice and video call options
- Maintain backward compatibility with prefill-only method
- Reuse existing Options.Kind enum for type safety
- Update documentation to clarify prefill vs auto-start behavior

lukelabonte committed:
Remove unused dialerTargetNumberPath constant

lukelabonte committed:
Replace magic strings with named constants

- Add constants for scheme name, base path, and query parameter keys
- Build doximityScheme from doximityName for single source of truth
- Update URLComponents construction to use constants throughout
- Simplify Options.path by removing local variable

lukelabonte committed:
Refactor API to use explicit methods instead of Options parameter

- Replace single method with options parameter with three explicit methods
- Add dialPhoneNumber() for prefill (backward compatible)
- Add startVoiceCall() for auto-starting voice calls
- Add startVideoCall() for auto-starting video calls
- Make Options enum internal, used only by private dial() implementation
- Update Objective-C bridge to match new API without enum bridging issues
- Fix URL construction to avoid triple-slash issue in scheme
- Add comprehensive documentation for private dial() method

lukelabonte committed:
Migrate tests to Swift Testing and add comprehensive coverage

- Replace XCTest with Swift Testing framework
- Fix existing test bugs (logic was backwards for installed/not installed)
- Add tests for all three dialing methods (prefill, voice, video)
- Add tests for URL encoding special characters
- Add tests for Options enum path generation and equatability
- Add tests for icon methods (regular and template modes)
- Add tests for Objective-C bridge icon methods
- Remove Objective-C bridge dial tests (can't mock singleton)

lukelabonte committed:
Add test injection for singleton and Objective-C bridge tests

- Add DEBUG-only setTestApplication() method for test injection
- Make singleton testable by using static test override in DEBUG builds
- Add tests for Objective-C bridge methods (prefill, voice, video)
- Serialize test suite to prevent race conditions with shared state
- Enable comprehensive testing of both Swift and Objective-C APIs

lukelabonte committed:
Simplify singleton test injection with instance replacement

- Replace computed property approach with direct singleton replacement
- Make shared instance public private(set) for external read-only access
- Add setSharedInstance() for DEBUG-only singleton replacement
- Centralize test setup in suite init() to eliminate per-test boilerplate
- Use shared singleton in all tests for more realistic testing
- Remove redundant nil checks on non-optional return values

lukelabonte committed:
Refactor tests into nested class hierarchy for better organization

- Change test suite from struct to class to enable subclassing
- Organize tests into nested classes: CoreFunctionality, ObjCBridge, Icon, Options
- Simplify test names by removing redundant prefixes (context now provided by class names)
- Each nested class inherits application instance from parent via subclassing
- Maintains serialized execution to prevent race conditions with shared singleton

lukelabonte committed:
Use backtick syntax for test names to eliminate redundancy

- Replace @Test("description") with @Test and backtick function names
- Apply backtick syntax to suite and class names for consistency
- Remove redundant suite title parameter, using class name instead
- Leverages Swift feature allowing spaces in backtick-escaped identifiers

lukelabonte committed:
Enhance SDK with installation check and comprehensive documentation

- Add public isDoximityInstalled property to check app installation status
- Include Objective-C bridge for isDoximityInstalled
- Document startVoiceCall and startVideoCall methods in README
- Add prominent Info.plist configuration warning section
- Reorganize README: requirements and integration before usage examples
- Add error handling examples for icon methods
- Fix bundle name reference: CallWithDoxDialer.bundle → DoximityDialerSDK.bundle
- Fix iconFromPackage to return nil instead of empty UIImage
- Add comprehensive tests for installation checking (Swift and Objective-C)
- Clarify phone number format support applies to all methods
- Add import statements to code examples
- Convert isDoximityInstalled to computed property for idiomatic Swift

lukelabonte committed:
Improve test names and reorganize test suite structure

- Shorten test names by removing redundant context (e.g., "when Doximity is installed")
- Reorganize test classes: Core Functionality, Installation Check, Icon, Options, Objective C Bridge
- Move Objective C Bridge class to bottom of file for better logical flow
- Maintain clarity while reducing verbosity in test descriptions

lukelabonte committed:
Added an example app

lukelabonte committed:
Move isDoximityInstalled to public API section

Relocate isDoximityInstalled from private extension to public API section
for better code organization and discoverability. The property remains
public and functionally identical - this is purely a structural improvement
to group public APIs together.

lukelabonte committed:
Update App Store redirect URL to use Singular deep link

Replace AppsFlyer URL with Singular deep link for App Store redirects.
The new URL provides better tracking and directs users to the Doximity
Dialer home page when the app is not installed.

- Remove unused appsFlyerID constant
- Update doximityAppStoreURL to use Singular link format
- Update corresponding test expectation

lukelabonte commented on Sources/DoximityDialerSDK/DoximityDialer.swift:
```diff
@@ -4,40 +4,133 @@ import UIKit
 public struct DoximityDialer {
     /// Singleton instance of `DoximityDialer`.
-    public static let shared = DoximityDialer()
+    public private(set) static var shared = DoximityDialer()
```
This was changed to facilitate testing, specifically the Obj-C bridging.

lukelabonte commented on Sources/DoximityDialerSDK/DoximityDialer.swift:
```diff
@@ -58,22 +156,17 @@ private extension DoximityDialer {
     /// Doximity icon asset from the swift package resources
     var iconFromPackage: UIImage? {
     #if SWIFT_PACKAGE
-        guard
-            let image = UIImage(
-                named: "icon",
-                in: .module,
-                compatibleWith: nil
-            )
-        else {
-            return UIImage()
-        }
-        return image
+        return UIImage(
+            named: "icon",
+            in: .module,
+            compatibleWith: nil
+        )
```
Now if we fail to get the image, the call gets back `nil` instead of a blank image.

lukelabonte commented on Sources/DoximityDialerSDK/DoximityDialer.swift:
```diff
@@ -58,22 +156,17 @@ private extension DoximityDialer {
     /// Doximity icon asset from the swift package resources
     var iconFromPackage: UIImage? {
     #if SWIFT_PACKAGE
-        guard
-            let image = UIImage(
-                named: "icon",
-                in: .module,
-                compatibleWith: nil
-            )
-        else {
-            return UIImage()
-        }
-        return image
+        return UIImage(
+            named: "icon",
+            in: .module,
+            compatibleWith: nil
+        )
     #else
         return nil
     #endif
     }
 
-    /// Doximity icon asset from the `CallWithDoxDialer.bundle` for manual integration
+    /// Doximity icon asset from the `DoximityDialerSDK.bundle` for manual integration
```
This was missed in a previous renaming.

lukelabonte commented on Sources/DoximityDialerSDK/DoximityDialer.swift:
```diff
@@ -85,23 +178,17 @@ private extension DoximityDialer {
         return doximityIcon
     }
 
-    /// Checks if the Doximity app is installed on the device.
-    var isDoximityInstalled: Bool {
-        guard let doximitySchemeURL = URL(string: Constants.doximityScheme) else {
-            return false
-        }
-        return application.canOpenURL(doximitySchemeURL)
-    }
-
     /// URL for the Doximity app in the App Store, including app-specific parameters for tracking.
     var doximityAppStoreURL: URL? {
         let appName: String = Bundle.main.infoDictionary?["CFBundleDisplayName"] as? String ?? "Unknown"
         let appIdentifyingName = appName.replacingOccurrences(of: " ", with: "-")
-        return URL(string: "https://app.appsflyer.com/\(Constants.appsFlyerID)?pid=third_party_app&c=\(appIdentifyingName)")
+
+        return URL(string: "https://doximity.sng.link/Az2j3/c17w?_dl=https://www.doximity.com/dialer/home&_smtype=3&pid=third_party_app&c=\(appIdentifyingName)")
```
Since we stopped using AppsFlyer, swapped this to a Singular link.

lukelabonte commented on README.md:
```diff
@@ -13,69 +13,150 @@ Doximity lets healthcare professionals make phone calls to patients while on the
 
 This µLibrary lets 3rd-party apps easily initiate calls through the Doximity app.
 
-## Other Platforms
+**Requirements:** iOS 15.0+
 
-* [Android](https://github.com/doximity/android-dialer-call-lib)
+## ⚠️ Required Configuration
+
+**Before using this SDK, you MUST configure your app's Info.plist:**
```
If you want to see how Github will render this, you can click the 3 dots on the upper right of this file, and then click on `View file`.

Otherwise [this](https://github.com/doximity/CallWithDoxDialer/blob/c7653cc755782f3777f6670df8cd83af876f7c77/README.md) is a direct link, but it may be stable if any changes get pushed to this file.

cpatterson-dox commented on Examples/DoximityDialerExample/README.md:
```diff
@@ -0,0 +1,96 @@
+✅ Auto-start voice calls
+✅ Auto-start video calls
+✅ Icon loading and display
+✅ Proper Info.plist configuration
```
_nitpick_: This checklist flows into a single line of text when rendered on Github. Insert `<br>` line breaks to split them up?

lukelabonte replied:
Address in 1f60f8b605d36dfc3d41012a4098809bdde9c3ee.

cpatterson-dox reviewed (approved):
🤩 STELLAR work, @lukelabonte ! Nice improvements via the sample app, unit test coverage and multi-language support! Our partners will appreciate it!

Just one minor formatting nitpick on the README.

(Hopefully we can get Epic to use this now as well.)

cpatterson-dox commented:
@lukelabonte Ran through the QA developer checklist, and found 3 issues with the example app. 

1. If the app is already running and you then delete the Doximity app, it does not update its installation badge status on the home screen until killed and relaunched. 
2. The SwiftUI Example view behaves differently when the app is not installed. On the other views, it disables the main action buttons and shows an additional "Install Doximity" button. The SwiftUI screen does not do that. It keeps all buttons enabled and navigates to the App Store link for each one. Is that intentional? Additionally, the status text is a bit different.
3. The SwiftUI Example does not display an error alert when you tap a button with no phone number entered.

| SwiftUI Example | Other Examples |
|--------|--------|
| <img width="600" alt="SwiftUI Example" src="https://github.com/user-attachments/assets/0be01265-b8d7-4031-8ae6-549b0aea4e30" /> | <img width="600" alt="Swift Example" src="https://github.com/user-attachments/assets/0fce994e-f1d1-4fcf-8cfb-85b0e92bc44f" /> |

lukelabonte committed:
Code review feedback

lukelabonte committed:
Replace backtick syntax in tests with camelCase naming

- Replace all backtick-wrapped test class names with camelCase
- Replace all backtick-wrapped test method names with test prefix convention
- Fixes compatibility issues with iOS 16.3 and 16.4 CI builds

cpatterson-dox reviewed (approved):
Looks great!
